lxd/metrics/api_rates.go

```diff
@@ -0,0 +1,167 @@
+package metrics
+
+import (
+	"net/url"
+	"strings"
+	"sync/atomic"
+
+	"github.com/canonical/lxd/shared/entity"
+	"github.com/canonical/lxd/shared/version"
+)
+
+// requestResult represents a completed request status category.
+type requestResult string
+
+const (
+	errorServer requestResult = "error_server"
+	errorClient requestResult = "error_client"
+	success     requestResult = "succeeded"
+)
+
+// PossibleRequestResults contains all the possible request result categories.
+var PossibleRequestResults = []requestResult{
+	errorServer,
+	errorClient,
+	success,
+}
+
+// resolveReqResult categorizes a status code between serverError, clientError and success.
+// The status code could be coming from an asynchronous operation or a request response.
+func resolveReqResult(statusCode int, asyncOperation bool) requestResult {
+	// Status codes from operations follow a different pattern.
+	if asyncOperation {
+		// If status is not 200(succeeded) nor 401(cancelled), it is considered a failure.
+		if statusCode != 200 && statusCode != 401 {
+			return errorServer
+		}
+
+		return success
+	}
+
+	// 4* codes are considered client errors on HTTP.
+	if statusCode >= 400 && statusCode < 500 {
+		return errorClient
+	}
+
+	// 2* codes are considered successes on HTTP, 300 are for forwarded requests
+	// that are included as successes for the purpose of the metrics.
+	if statusCode >= 200 && statusCode < 400 {
+		return success
+	}
+
+	return errorServer
+}
+
+// resolveEntityType derives the main entity type an endpoint is opertaing on from its URL.
+// This is used to categorize completed requests in internal metrics.
+func resolveEntityType(url url.URL) entity.Type {
```

Comment:
Shall we have some unit tests for this?

Reply:
Makes sense to me. But perhaps we should wait on the entity type refactor since it could affect the behavior of this function.

Reply:
Yes i think we should wait.